### PR TITLE
[Deprecated by #27] Observe keyboard presentation and dismissal

### PR DESCRIPTION
**Pull request by aasatt**
For use when presented controller is presented as a popup. Use case for
when user has a text field and doesn’t want the keyboard to cover the
bottom part of the presented view which would normally contain some
kind of button.

The view also won't dismiss on the tap of the container view when the keyboard is being shown. Instead, it will simply dismiss the keyboard. Added 'PresentrKeyboardDelegate' that has 'shouldDismissKeyboard' to which the presented view controller can conform to. This will notify that controller of the tap on the container so it can dismiss the keyboard

Added 'KeyboardTranslationType' with '.MoveUp', '.Compress', '.None'. Move up will simply translate the view up on the y axis. Compress compresses the view between the keyboard and the status bar.

### CODE REVIEW

**Comment by tbaranes** [comment]
Since I'm also interesting in that feature, I take a look of the usage 😆 
I just have a question: why `observeKeyboard` is public? Its state should be directly related to `keyboardTranslationType`, if the value is `.Compress` or `.MoveUp`, then the keyboard must be observed, whereas if the value is `.None`, the keyboard won't be observed.

File: Presentr/Presentr.swift [hunk 1]
```diff
@@ -38,6 +39,12 @@ public class Presentr: NSObject {
     /// Should the presented controller dismiss on background tap. Default is true.
     public var dismissOnTap = true
     
+    /// Should the presenter controller observe the showing and dismissal of the keyboard. Default is false. Only avaiable Popup Presentation Type. for Use this when you have a text field or text view in the presented controller to use taps on the outside of the controller to dismiss the keyboard instead of dismissing the controller. Also translates the view controller based on the keyboard offset if part of the view would be covered by the keyboard.
```
Comment (tbaranes) [inline Presentr/Presentr.swift]: Typo: `avaiable` -> `available`

**Comment by aasatt** [comment]
@tbaranes Good point. #16 originally had just the MoveUp translation but I needed to support Compress for a different use case. I guess adding the TranslationType made 'observeKeyboard' redundant. You're right to point out there virtually no need for it now.

**Comment by aasatt** [comment]
@tbaranes Oh. Just remember why it's relevant. You can observe keyboard without the translation of the view if you just want the tap on the container view to not dismiss the keyboard.

**Comment by tbaranes** [comment]
Nice feature 🤔In that case, I think you should rename it. Having these both variables will be definitely confusing!

**Comment by danlozano** [comment]
Nice work guys! Sorry I'm late I was busy but finally got some time to look this over. This will probably be a common use case so it's important we get it right.

Looked over the code, at first glance looks good, just have a couple of points to be looked at. Discussion/comments from both are welcome :)

Starting in the **KeyboardTranslation.swift** file

### 1st Point

```swift
public protocol PresentrKeyboardDelegate {
```

- I think this could be called **PresentrDelegate**, which is a protocol that the Presented View Controller can conform to, if they want to make any decision or be alerted of any actions. This way any other delegate methods we want to add later on could be added to this protocol.

We could have these two methods for now.
```swift
func keyboardShouldPreventDismissal -> Bool{
    return true
}
func chromeViewTapped{
    // They can perform any extra actions when the chrome view is tapped, before the VC is dismissed. 
//If they return true from the keyboardShouldPreventDismissal method, then this is where they would dismiss the keyboard.
}
```

This way, on @tbaranes point. The observeKeyboard public property does even not have to exist. We get this information in the **PresentrController** through private computed properties. If the presented view controller cares about the keyboard dismissal OR if the keyboardTranslationType is different than .None, then we care about the keyboard notifications, if not we don't.

```swift

private var shouldPresentedPreventDismissal: Bool{
    let presented = presentedViewController as? PresentrDelegate
    return presented?.keyboardShouldPreventDismissal
}

private var shouldObserveKeyboard: Bool{
    return shouldPresentedPreventDismissal ||
    ((keyboardTranslationType != .None) && presentationType == .Popup)
}
```

### 2nd Point

```swift
public static func getTranslationFrame (type : KeyboardTranslationType, keyboardFrame : CGRect, presentedFrame : CGRect) -> CGRect {
```
I think this method should **not** be **static.** It's a method that makes calculations and returns a value based on the **enum** case. So this should be an instance method, that way you no longer have to pass in the KeyboardTranslationType as an argument, it could just switch on Self. It could look something like this.

```swift
public func getTranslationFrame(keyboardFrame : CGRect, presentedFrame : CGRect) -> CGRect {
    switch self{
        case .MoveUp
        case .Compress
        etc...
    }
}
```

### 3rd Point

Why is this only available when the PresentationType is .Popup ? Should this be available in other types? 

What do you guys think?

Happy friday ! :)

**Comment by aasatt** [comment]
### 1st Point 
I agree. This will lay down the foundation for additional features.

### 2nd Point 
Yup. This is probably the best design. 

### 3rd Point 
It doesn't have to be. To be honest, this was the only use case I had it for and I was on a time constraint with the project I was working on so I didn't have time to test it with the other presentation types. 


It's good to have another set of eye's looking over this. I merely wanted to provide a starting point to implement this.  Great job Presentr so far! I am really enjoying it.

**Comment by danlozano** [comment]
Yeah, that's what this is for. To provide a starting point, look at it together, discuss it, iterate over it, etc... :)

Well I think for now we can start with .Popup, and get the architecture all worked out. 

Later on I can probably get started contributing some code and we can start messing with other PresentationType's , and see if there's any extra handling we need to do depending on the type/size.

I think it's looking pretty good! Thanks for contributing, this is the first time I've open sourced something, the feedback and contribution has been great so far

:+1:

**Comment by aasatt** [comment]
I have time to work on this now. Couple questions:

1. Should `PresentrKeyboardDelegate`'s methods be optional?

2. `keyboardShouldPreventDismissal` could be `presentrShouldDismiss() -> Bool` which would default to true. 

     - This would behave like a `textFieldShouldReturn`.

     -  This is the time for a delegate to perform any actions on dismissal or prevent the dismissal entirely. Would this also replace the `chromeViewTapped`?

 The use case for this would be the user wants to know if the chrome view was tapped but needs to verify some properties before allowing it to dismiss. That way it doesn't have to be exclusive to the keyboard. 

Let me know what you think and if we should keep your originally proposed methods for now or not.

**Comment by danlozano** [comment]
Nice!

1. I think they **should** be optional, that's why in the PresentrController we check to see if the presented view controller conforms to it or not.

2. Oh ok. My first thought was going to be that the function name should include "keyboard" because it was only related to the keyboard, but if it is going to be generalized then I agree with your naming _**presentrShouldDismiss**_

So, we could use that delegate method so the user can do anything before the dismissal, AND prevent it, if they choose to.

This would make this Presentr property redundant.

```swift
    /// Should the presented controller dismiss on background tap. Default is true.
    public var dismissOnTap = true
```

But, this design would make it the user's responsibility to know when the keyboard is showing or not, and IF it is showing, it is up to them to prevent the dismissal in the _**presentrShouldDismiss**_ delegate method. 

The reason being because since now we are calling **_presentrShouldDismiss_** whenever someone taps on the chrome view, not just only when the keyboard was showing, in the old "design".


On second thought, I think I don't like this because we are possibly already listening to the keyboard notifications (for doing the translation), so it's redundant for the user to have to listen to the keyboard as well in the presented view controller. A solution that comes to mind is having the delegate method be like so:

```swift
func presentrShouldDismiss(keyboardShowing: Bool) -> Bool{
   ...
   return true
}
```

That way, with that delegate method we cover all cases. If they tap on the chrome view, and the keyboard is not showing, or if the keyboard is showing, we give them the information they need. It is up to them to decide to dismiss it or not, and do something before.

We could then maybe remove these two methods:

```swift
private var shouldPresentedPreventDismissal: Bool{
    let presented = presentedViewController as? PresentrDelegate
    return presented?.keyboardShouldPreventDismissal
}

private var shouldObserveKeyboard: Bool{
    return shouldPresentedPreventDismissal ||
    ((keyboardTranslationType != .None) && presentationType == .Popup)
}
```

Since we would always want to listen to the keyboard.

What do you think?

**Comment by danlozano** [comment]
Correction,

We would still need both methods , but they would be like so:

```swift

private var presentedConformsToProtocol: Bool{
    return presentedViewController is PresentrDelegate
}

private var shouldObserveKeyboard: Bool{
    return presentedConformsToProtocol ||
    ((keyboardTranslationType != .None) && presentationType == .Popup)
}
```

That way, we only observe the keyboard **IF** they chose a keyboardTranslationType != .None ; or if they conform to the protocol, which means, they want to know when we want to dismiss the controller, and we need to send the keyboard showing bool.

If they don't implement the protocol, and choose no translationType, there is no reason for us to listen to the keyboard.

**Comment by aasatt** [comment]
I finished up the changes and this time on a separate branch off of the Develop. That way you can decide to merge it how you want. Instead of trying to revert and rebase this pull request, would it be appropriate to create yet another pull request?

**Comment by danlozano** [comment]
Awesome!

Yeah, I guess another PR would be best in this case 👍

**Comment by gabrielPeart** [comment]
@aasatt I don't have time to check it right now, but could you confirm if is the same or related to #27 ?

**Comment by aasatt** [comment]
@gabrielPeart This was my first attempt. Taking the directive of @danlozano I created #27 which is on it's own branch off of develop where this one was just on the master. I didn't want to have to screw around with the git to rebase it and everything. This one, #18 can be closed.

**Comment by gabrielPeart** [comment]
@aasatt That is what I presumed. Could you close it?
Thank you!

**Comment by aasatt** [comment]
@gabrielPeart wow it took me way longer than I would like to admit to see the big close button right below this comment box... I think it's been like 16 days before I first tried to figure it out 😅

**Comment by gabrielPeart** [comment]
@aasatt Cool :)
Thank you!

**Pull request closed without merging**